src/main/java/se/liu/ida/hefquin/engine/query/impl/QueryPatternUtils.java

```diff
@@ -182,6 +177,88 @@ else if ( p instanceof GenericSPARQLGraphPatternImpl2 ) {
 		}
 	}
 
+	/**
+	 * Returns the set of all triple patterns that occur in the given graph pattern.
+	 */
+	public static Set<TriplePattern> getTPsInPattern( final SPARQLGraphPattern queryPattern ) {
+		final Set<TriplePattern> tps = new HashSet<>();
+		if ( queryPattern instanceof TriplePattern ) {
+			tps.add( (TriplePattern) queryPattern );
+		}
+		else if ( queryPattern instanceof BGP ) {
+			tps.addAll( ((BGP) queryPattern).getTriplePatterns() );
+		}
+		else if ( queryPattern instanceof SPARQLGroupPattern ) {
+			final SPARQLGroupPattern gp = (SPARQLGroupPattern) queryPattern;
+			for ( int i = 0; i < gp.getNumberOfSubPatterns(); i++ ) {
+				tps.addAll( getTPsInPattern(gp.getSubPatterns(i)) );
+			}
+		}
+		else if ( queryPattern instanceof SPARQLUnionPattern ) {
+			final SPARQLUnionPattern up = (SPARQLUnionPattern) queryPattern;
+			for ( int i = 0; i < up.getNumberOfSubPatterns(); i++ ) {
+				tps.addAll( getTPsInPattern( up.getSubPatterns(i) ) );
+			}
+		}
+		else if ( queryPattern instanceof GenericSPARQLGraphPatternImpl1 ) {
+			final Element element = ( (GenericSPARQLGraphPatternImpl1) queryPattern ).asJenaElement();
+			return getTPsInPattern(element);
+		}
+		else if ( queryPattern instanceof GenericSPARQLGraphPatternImpl2 ) {
+			final Op jenaOp = ( (GenericSPARQLGraphPatternImpl2) queryPattern ).asJenaOp();
+			return getTPsInPattern(jenaOp);
+		}
```

Comment:
Since these two cases do not need the creation of the HashSet `tps`, can you please move them before the (currently) first line of this function; i.e., before the creation of `tps`. You can also have both of them just be if-blocks (i.e., not if-else) because each of them immediately returns something.